File: models/client_model.py
from pydantic import BaseModel, field_validator
from typing import Annotated
from sqlalchemy import Table, Column, DateTime, Integer, String, Boolean
from utils.database import metadata
from datetime import datetime as dt
from schemas.client_schema import ClientBase

# Define the clients table
clients_table = Table(
    "clients",
    metadata,
    Column("id", Integer, primary_key=True, autoincrement=True),
    Column("prefix", String(10), nullable=False, default=""),
    Column("name", String(255), nullable=False),
    Column("address", String(255), nullable=False),
    Column("city", String(100), nullable=False),
    Column("province", String(100), nullable=False),
    Column("phoneNumber", String(20), nullable=False),
    Column("email", String(255), nullable=True),
    Column("npwp", String(50), nullable=True),
    Column("createdAt", DateTime(), nullable=False, default=dt.now),
    Column("createdBy", Integer, nullable=False),
    Column("updatedAt", DateTime(), nullable=True, default=None),
    Column("updatedBy", Integer, nullable=True, default=None),
    Column("isDelete", Boolean, nullable=False, default=False),
    Column("deletedAt", DateTime(), nullable=True, default=None),
    Column("deletedBy", Integer, nullable=True, default=None),
)

File: src/models/client.model.ts
import type { IClientCreate, IClientOutput, IClientUpdate, IClientView } from "../interfaces/client.interfaces"
import type { IUserView } from "../interfaces/user.interfaces"

export class ClientModel {
  static toOutput(client: any, createdBy: IUserView, updatedBy: IUserView | null): IClientOutput {
    return {
      id: client.id,
      prefix: client.prefix,
      name: client.name,
      address: client.address,
      city: client.city,
      province: client.province,
      phoneNumber: client.phoneNumber,
      email: client.email,
      npwp: client.npwp,
      createdAt: client.createdAt,
      updatedAt: client.updatedAt ?? null,
      createdBy,
      updatedBy,
    }
  }

  static toView(client: any): IClientView {
    return {
      id: client.id,
      prefix: client.prefix,
      name: client.name,
      city: client.city,
      province: client.province,
      phoneNumber: client.phoneNumber,
    }
  }

  static createPayload(data: IClientCreate) {
    return {
      ...data,
      prefix: data.prefix ?? "",
    }
  }

  static updatePayload(data: IClientUpdate) {
    return {
      ...data,
    }
  }
}
